### PR TITLE
#97 Implemented delayed validation by introducing delayedValidation o…

### PR DESCRIPTION
**Pull request by andkorsh**
#97 Implemented delayed validation by introducing delayedValidation option.

The logic is a little bit more complicated than switching validation event to blur as initially discussed in the issue. This implementation doesn't do any validation while the user typing for the first time (i.e. until the first 'change' or 'submit') but once the validation had failed and user focused back into the input, validation switches back to "input" event to remove the error as soon as the data is valid.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1068,6 +1069,15 @@ export class ValidationService {
...
             let validate = this.validators[uid];
             if (!validate) return true;
+            
+            if (
+                this.options.delayedValidation &&
+                event && event.type === 'input' &&
+                !input.classList.contains(this.ValidationInputCssClassName)
+            ) {
+                // When delayedValidation=true, "input" only takes it back to valid. "Change" can make it invalid.
+                return;
```
Comment (haacked) [inline src/index.ts:1079]: This needs to return `true` or `false`.

**Comment by haacked** [comment]
Do you mind adding a sample page with delayed validation enabled so it's easy to test the behavior?

**Comment by andkorsh** [comment]
@haacked sure, thanks for looking at this. Fixed the return statement and added a sample. Please have a look when the time allows. We'd love to see it in the package to use it in our app.

Also below is a gif demonstrating the behavior of new delayed validation. Idea is, it doesn't bother you while you still typing, but only shows up when you finished and focused out/submitted. And if the field is errored and you are back in it to fix it, it will turn back valid as soon as the data becomes valid.

delayed_validation

In fact, the original lib at https://jqueryvalidation.org/files/demo/ behaves very similarly:

delayed_validation (original)

**Comment by haacked** [comment]
Oh yeah, the current behavior is annoying. It seems to me the behavior in this PR should be the default as it seems to align with the jquery validation behavior. @dahlbyk any thoughts?

**Comment by andkorsh** [comment]
@haacked If we choose to make this behavior default, I think it would be better to remove the `delayedValidation` option altogether, leaving users with `data-val-event="input"` should they require the old behavior. Also we'd need to refactor the below lines and get rid of nested ternary. When I implemented this PR, I tried not to touch any of the existing behaviors and did not fix the bug where `data-val-event` value isn't considered at all. But if we do choose to introduce a change in default behavior, fixing the bug would be logical.

https://github.com/haacked/aspnet-client-validation/blob/a59c47aebbc94cb3e930c3685c6865c8963a183a/src/index.ts#L1102-L1104

 Please let me know what you decide and I can take care of it.

cheers,
Andrew

**Comment by haacked** [comment]
@andkorsh I agree with your plan. Let’s make it so.

**Comment by andkorsh** [comment]
@haacked I have updated the PR now and created a "Immediate Validation" sample page that comparing the new default behavior with the old behavior which can now be achieved using `data-val-event="input"`.

cheers!

File: src/index.ts [hunk 2]
```diff
@@ -1068,6 +1068,15 @@ export class ValidationService {
         const cb: ValidationEventCallback = async (event, callback) => {
             let validate = this.validators[uid];
             if (!validate) return true;
+            
+            if (
+                !input.dataset.valEvent &&
+                event && event.type === 'input' &&
+                !input.classList.contains(this.ValidationInputCssClassName)
+            ) {
```
Comment (haacked) [inline src/index.ts:1076]: Minor nit, but I prefer the logical operators in a multi-line conditional to align on the left. Makes it easier to see all the component conditionals.

```suggestion
            if (!input.dataset.valEvent
                && event
                && event.type === 'input'
                && !input.classList.contains(this.ValidationInputCssClassName)
            ) {
```

**Review by haacked — COMMENTED** [review]
Looks good! I need to test it before I approve it.

**Comment by andkorsh** [comment]
Thank you! And appreciate you creating this great package, hope Microsoft will use it and remove the jquery dependency from identity templates one day.

**Comment by andkorsh** [comment]
It is still a drop in replacement and behavior now aligns more with how the
original library behaves.

cheers

**Pull request merged**